### ISSUE TITLE
Suggestion: Reasoning for non-rules

### ISSUE THREAD

**Issue by eedahl** [isocpp/CppCoreGuidelines#632]
The non-rules is a great section (to me as a novice/intermediary) I wish was expanded with the same level of reasoning as the rest of the document.

When bad advice is ubiquitous, it is arguably as important to get a good reason not to follow whatever bad advice you've heard as it is to follow good advice given good reasons.

Especially when learning good and bad advice are often indistinguishable, and trying to follow advice until you've either seen why it's clever or find a very good reason not to is by and large a sound learning strategy.

In other words: giving negative advice to counter bad advice is still good advice, and should have a good reason and/or examples of good practice violating the non-rules.

> Non-rule summary:
> 
> all declarations on top of function
> single-return rule
> no exceptions
> one class per source file
> two-phase initialization
> goto exit

**Comment by BjarneStroustrup** [comment]
Correct. We just haven't gotten around to expanding that section.

**Comment by BjarneStroustrup** [comment]
Some discussion and examples added.
